jgrapht-core/src/main/java/org/jgrapht/alg/flow/PushRelabelMaximumFlow.java

```diff
@@ -0,0 +1,388 @@
+ *
+ * (C) Copyright 2003-2008, by Barak Naveh and Contributors.
+ *
+ * This program and the accompanying materials are dual-licensed under
+ * either
+ *
+ * (a) the terms of the GNU Lesser General Public License version 2.1
+ * as published by the Free Software Foundation, or (at your option) any
+ * later version.
+ *
+ * or (per the licensee's choosing)
+ *
+ * (b) the terms of the Eclipse Public License v1.0 as published by
+ * the Eclipse Foundation.
+ */
+/* -----------------
+ * PushRelabelMaximumFlow.java
+ * -----------------
+ * (C) Copyright 2015-2015, by Alexey Kudinkin and Contributors.
+ *
+ * Original Author:  Alexey Kudinkin
+ * Contributor(s):
+ *
+ * $Id$
+ *
+ * Changes
+ * -------
+ */
+package org.jgrapht.alg.flow;
+
+import org.jgrapht.DirectedGraph;
+import org.jgrapht.alg.util.Extension.ExtensionFactory;
+import org.jgrapht.alg.util.Pair;
+
+import java.util.*;
+
+/**
+ * <p><a href="https://en.wikipedia.org/wiki/Push%E2%80%93relabel_maximum_flow_algorithm">
+ * Push-relabel maximum flow</a> algorithm designed by Andrew V. Goldberg and
+ * Robert Tarjan. Current implementation complexity upper-bound is O(V^3). For more
+ * details see: <i>"A new approach to the maximum flow problem"</i> by Andrew V. Goldberg
+ * and Robert Tarjan <i>STOC '86: Proceedings of the eighteenth annual ACM symposium on
+ * Theory of computing</i></p>
+ *
+ * @author Alexey Kudinkin
+ */
+public class PushRelabelMaximumFlow<V, E> extends MaximumFlowAlgorithmBase<V,E> {
+
+    private static final boolean LABEL_PRUNE_ENABLED = true;
+
+    private DirectedGraph<V, E> network;
+
+    private final ExtensionFactory<VertexExtension> vertexExtensionsFactory;
+    private final ExtensionFactory<EdgeExtension>   edgeExtensionsFactory;
+
+    // Label pruning
+
+    private Map<Integer, Integer> labeling;
+
+    boolean flowBack;
+
+    // Diagnostic
+
+    private static final boolean DIAGNOSTIC_ENABLED = false;
+
+    private class PushRelabelDiagnostic {
+
+        // Discharges
+        Map<Pair<V, V>, Integer> discharges = new HashMap<Pair<V, V>, Integer>();
+        long dischargesCounter = 0;
+
+        private void incrementDischarges(EdgeExtension ex) {
+            Pair<V, V> p = Pair.of(ex.getSource().prototype, ex.getTarget().prototype);
+            if (!discharges.containsKey(p))
+                discharges.put(p, 0);
+            discharges.put(p, discharges.get(p) + 1);
+
+            dischargesCounter++;
+        }
+
+        // Relabels
+        Map<Pair<Integer, Integer>, Integer> relabels = new HashMap<Pair<Integer, Integer>, Integer>();
+        long relabelsCounter = 0;
+
+        private void incrementRelabels(int from, int to) {
+            Pair<Integer, Integer> p = Pair.of(from, to);
+            if (!relabels.containsKey(p))
+                relabels.put(p, 0);
+            relabels.put(p, relabels.get(p) + 1);
+
+            relabelsCounter++;
+        }
+
+        void dump() {
```

Comment:
Please document your functions. That makes it easier for others to extend/improve/fix issues in your code.